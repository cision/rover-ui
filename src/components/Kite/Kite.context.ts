import { createContext } from 'react';

type ContextProps = {
  canBeDismissed?: boolean;
  onClose?: () => void;
};

export default createContext<ContextProps>({});
