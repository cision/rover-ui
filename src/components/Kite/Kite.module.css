.Kite {
  z-index: var(--rvr-zindex-kite);
  visibility: hidden;
  box-sizing: border-box;
  position: fixed;
  top: var(--rvr-space-md);
  right: var(--rvr-space-md);
  transition: transform .6s ease-in-out;
  animation: slide-in-right .7s;
  display: flex;
  flex-flow: row nowrap;
  align-items: flex-start;
  background-color: var(--rvr-white);
  width: 342px;
  box-shadow: 0px 5px 15px rgba(0, 0, 0, 0.2);
  border-radius: 4px;
  padding: 16px;
  font-size: var(--rvr-font-size-md);
  color: var(--rvr-color-font-dark);
}

.Kite > * {
  margin-left: 16px;
}

.Kite > :first-child {
  margin-left: 0;
}

.visible {
  visibility: visible;
}

.icon {
  flex: 0 0 auto;
  align-self: flex-start;
}

.body {
  flex: 1 1 auto;
  min-width: 0;
}

.dismissButton {
  flex: 0 0 auto;
  margin-right: -16px;
  margin-top: -12px;
  margin-bottom: -12px;
  align-self: flex-start;
}

.header {
  font-weight: var(--rvr-font-weight-bold);
}

.header + .content {
  margin-top: 8px;
}

.title {
  flex-grow: 2;
}

.enterDone {
  opacity: 1;
  transform: translateX(0);
}

.exit {
  opacity: 0;
  transform: translateX(-200px);
}

.top-right {
  top: 12px;
  right: 12px;
  transition: all .6s ease-in-out;
  transform: translateX(100%);
}




@keyframes slide-in-right {
  from {
    transform: translateX(100%);
  }

  to {
    transform: translateX(0);
  }
}
