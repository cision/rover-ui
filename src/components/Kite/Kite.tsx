import React, { useEffect, CSSProperties } from 'react';
import classNames from 'classnames';

import ReactDOM from 'react-dom';
import { CSSTransition } from 'react-transition-group';
import styles from './Kite.module.css';
import Button from '../Button';
import Icon from '../Icon';

interface KiteProps {
  children?: React.ReactNode;
  canBeDismissed?: boolean;
  className?: string;
  visible?: boolean;
  icon?: React.ReactNode;
  onClose?: () => void;
  style?: CSSProperties;
  ttl?: number;
}

type KiteIconProps = Parameters<typeof Icon>[0];
type KiteChildProps = Pick<KiteProps, 'children' | 'className'>;

type KiteType = React.FC<KiteProps> & {
  Header: React.FC<KiteChildProps>;
  Icon: React.FC<KiteIconProps>;
  Content: React.FC<KiteChildProps>;
};

const Kite: KiteType = ({
  canBeDismissed = true,
  children = undefined,
  className: passedClassName = '',
  visible = false,
  icon = undefined,
  onClose = () => {},
  ttl = undefined,
  ...passedProps
}) => {
  useEffect(() => {
    const interval = setInterval(() => {
      if (onClose && ttl) {
        onClose();
      }
    }, ttl);
    return () => {
      clearInterval(interval);
    };
  }, [onClose, ttl]);

  return ReactDOM.createPortal(
    <CSSTransition
      in={visible}
      unmountOnExit
      timeout={{ enter: 0, exit: 300 }}
      classNames={{
        enterDone: styles.enterDone,
        exit: styles.exit,
      }}
    >
      <div
        className={classNames(
          styles.Kite,
          { [styles.visible]: visible },
          passedClassName
        )}
        {...passedProps}
      >
        {icon && <div className={styles.icon}>{icon}</div>}
        <div className={styles.body}>{children}</div>
        {canBeDismissed && (
          <div className={styles.dismissButton}>
            <Button
              level="text"
              onClick={onClose}
              data-testid="kite-dismiss-button"
            >
              <Button.Addon>
                <Icon
                  height="20"
                  name="close"
                  style={{ display: 'block' }}
                  width="20"
                />
              </Button.Addon>
            </Button>
          </div>
        )}
      </div>
    </CSSTransition>,
    document.body
  );
};

const KiteIcon: React.FC<KiteIconProps> = ({ className, ...props }) => {
  return <Icon {...props} className={classNames(styles.icon, className)} />;
};

const KiteContent: React.FC<KiteChildProps> = ({ className, ...props }) => {
  return <div {...props} className={classNames(styles.content, className)} />;
};

const KiteHeader: React.FC<KiteChildProps> = ({ className, ...props }) => {
  return <div {...props} className={classNames(styles.header, className)} />;
};

Kite.Header = KiteHeader;
Kite.Icon = KiteIcon;
Kite.Content = KiteContent;
export default Kite;
