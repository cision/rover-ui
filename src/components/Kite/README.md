# \<Kite\>

**A configurable Kite component that automatically triggers onClose when the time to live interval elapses**

This component can be used as a toast that slides in from the top left to display information.

Below is an example structure of the `<Kite>`

```js
<Kite
  canBeDismissed
  icon={<Kite.Icon />}
  onClose={() => setVisible(false)}
  ttl={3000}
  visible={visible}
>
  <Kite.Header>Bold title</Kite.Header>
  <Kite.Body>Optional plain text / React node content</Kite.Body>
</Kite>
```

This will render a Kite with an icon, heading, and body content.

- The parent component controls visibility via the `onClose` callback and `visible` prop.
- If a `ttl` is provided, the component will call `onClose` that many ms after `visible` changes to `true`.
- `icon` and `children` can be anything React can render. Kites expert simple `Kite.Icon`, `Kite.Header`, and `Kite.Body` helpers. `Kite.Icon` uses the same interface as the base `Icon` component, and the heading and body provide some default spacing and font-weight variation.
- If `canBeDismissed` is false, the kite will not include a standard close button.
