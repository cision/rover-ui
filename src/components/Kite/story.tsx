import React, { useState } from 'react';
import { storiesOf } from '@storybook/react';

import Kite from '.';
import Readme from './README.md';
import { Wrap } from '../../stories/storybook-helpers';
import Button from '../Button';
import Icon from '../Icon';

function KiteExample() {
  const [isOpen, setIsOpen] = useState(false);

  const showKite = () => {
    setIsOpen(true);
  };

  const hideKite = () => {
    setIsOpen(false);
  };

  return (
    <>
      <div>
        <Button modifiers={['primary']} onClick={showKite}>
          Show Kite
        </Button>
      </div>
      <Kite
        title="Success Kite!"
        canBeDismissed
        visible={isOpen}
        onClose={hideKite}
        ttl={3000}
      >
        <Kite.KiteIcon>
          <Icon
            fill="green"
            height="20"
            name="check"
            style={{ display: 'block' }}
            width="20"
          />
        </Kite.KiteIcon>
      </Kite>
    </>
  );
}

function ErrorKiteExample() {
  const [isOpen, setIsOpen] = useState(false);

  const showKite = () => {
    setIsOpen(true);
  };

  const hideKite = () => {
    setIsOpen(false);
  };

  return (
    <>
      <div>
        <Button modifiers={['primary']} onClick={showKite}>
          Error Kite
        </Button>
      </div>
      <Kite
        title="Error Kite!"
        canBeDismissed
        visible={isOpen}
        onClose={hideKite}
        ttl={3000}
      >
        <Kite.KiteIcon>
          <Icon
            fill="red"
            height="20"
            name="warning"
            style={{ display: 'block' }}
            width="20"
          />
        </Kite.KiteIcon>
      </Kite>
    </>
  );
}

storiesOf('Galaxies/Kite', module)
  .addParameters({
    readme: {
      sidebar: Readme,
    },
  })
  .add(
    'Overview',
    () => (
      <Kite
        visible={false}
        title="Kite component!"
        canBeDismissed
        ttl={3000}
        onClose={() => {}}
      >
        <Kite.KiteIcon>
          <Icon
            fill="green"
            height="20"
            name="check"
            style={{ display: 'block' }}
            width="20"
          />
        </Kite.KiteIcon>
      </Kite>
    ),
    {
      info: {
        inline: true,
        source: true,
      },
    }
  )
  .add(
    'Examples',
    () => (
      <>
        <Wrap>
          <KiteExample />
        </Wrap>
        <Wrap>
          <ErrorKiteExample />
        </Wrap>
      </>
    ),
    {
      info: {
        inline: false,
        source: false,
      },
    }
  );
