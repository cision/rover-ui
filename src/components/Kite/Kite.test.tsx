import React from 'react';
import { render, screen } from '@testing-library/react';
import '@testing-library/jest-dom';

import userEvent from '@testing-library/user-event';
import Kite from './Kite';

const defaultProps = {
  visible: true,
  onClose: jest.fn(),
  ttl: 3000,
  canBeDismissed: true,
};

const renderKite = (props = defaultProps) =>
  render(
    <Kite
      {...props}
      icon={<Kite.Icon fill="red" height="20" name="warning" width="20" />}
    >
      <Kite.Header>Kite Title</Kite.Header>
      <Kite.Body>
        <p>Kite Content Goes here!</p>
      </Kite.Body>
    </Kite>
  );
describe('Kite', () => {
  it('renders correctly', () => {
    const { baseElement } = renderKite();
    expect(baseElement).toMatchSnapshot();
  });

  it("does not render when 'visible' prop is false", () => {
    render(<Kite visible={false} data-testid="Kite-Test" />);
    expect(screen.queryByTestId('Modal-Test')).not.toBeInTheDocument();
  });

  describe('Dismiss Button', () => {
    test.each`
      canBeDismissed | visible
      ${true}        | ${true}
      ${false}       | ${false}
    `(
      'when canBeDismissed = $canBeDismissed, the close button visible should be: $visible',
      ({ canBeDismissed }) => {
        render(
          <Kite {...defaultProps} canBeDismissed={canBeDismissed}>
            <Kite.Header>Title</Kite.Header>
          </Kite>
        );

        const dismissButton = screen.queryByTestId('kite-dismiss-button');
        if (canBeDismissed) {
          expect(dismissButton).toBeInTheDocument();
        } else {
          expect(dismissButton).not.toBeInTheDocument();
        }
      }
    );
  });

  describe('onClose callback', () => {
    it('calls onClose callback when dismiss button is clicked', async () => {
      render(
        <Kite {...defaultProps} onClose={defaultProps.onClose}>
          <Kite.Header>Title</Kite.Header>
        </Kite>
      );
      const dismissButton = screen.getByTestId('kite-dismiss-button');
      expect(dismissButton).toBeInTheDocument();

      userEvent.click(dismissButton);

      expect(defaultProps.onClose).toHaveBeenCalledTimes(1);
    });
  });
});
