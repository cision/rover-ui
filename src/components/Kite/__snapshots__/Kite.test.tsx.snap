// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Kite renders correctly 1`] = `
<body>
  <div />
  <div
    class="Kite visible"
  >
    <div
      class="icon"
    >
      <svg
        class="icon"
        fill="red"
        height="20"
        viewBox="0 0 24 24"
        width="20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M1 21h22L12 2 1 21zm12-3h-2v-2h2v2zm0-4h-2v-4h2v4z"
        />
      </svg>
    </div>
    <div
      class="body"
    >
      <div
        class="header"
      >
        Kite Title
      </div>
      <div
        class="content"
      >
        <p>
          Kite Content Goes here!
        </p>
      </div>
    </div>
    <div
      class="dismissButton"
    >
      <button
        class="Button text lg"
        data-testid="kite-dismiss-button"
        type="button"
      >
        <div
          class="Addon md"
        >
          <svg
            fill="currentColor"
            height="20"
            style="display: block;"
            viewBox="0 0 24 24"
            width="20"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
            />
          </svg>
        </div>
      </button>
    </div>
  </div>
</body>
`;
